Fix CLI connections import and migrate logic from secrets to Connection model

natanweinberger opened the pull request:
This is a bug fix for the CLI command `connections_import`. I originally added this CLI functionality in #15177. The bug was reported by @mudravrik [here](https://github.com/apache/airflow/pull/15177#issuecomment-822056108).

## Bug summary
Running `airflow connections import <filepath>` results in an AttributeError.

```
Traceback (most recent call last):
...
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/commands/connection_command.py", line 244, in connections_import
    _import_helper(args.file)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/cli/commands/connection_command.py", line 272, in _import_helper
    key: value for key, value in conn_values.items() if key in allowed_fields
AttributeError: 'Connection' object has no attribute 'items'
```

## Root cause
I called the existing function `load_connections_dict` to get the contents of the target file, thinking it returned a list of dictionaries. When I later looped through the list, I called the dictionary method `.items()` on each connection. However, each connection was actually stored into a `Connection` model instance rather than a dictionary, so the call I made to `.items()` resulted in an AttributeError.

## Solution
In taking a closer look at this issue, I realized the code block that contained the bug can be removed altogether. Having a list of `Connection` models to begin with (rather than dictionaries) eliminated the need to even examine the keys and values of the dictionaries. I had only done that to safely load the data into `Connection` models in the first place.

## Preventing future issues
To avoid future issues, I also changed how the tests work for providing sample data. Before, a dictionary of sample data was set as the return value of `load_connections_dict`, which actually returns an object of type `List[Connection]`. Now, that dictionary of sample data is set to be the return value of a lower level function, `_parse_secrets_file`, which does no heavy lifting outside of parsing a JSON, YAML or env file and returns a dictionary.

Bug originally introduced in #15177, which closed #9855.

mik-laj commented on airflow/cli/commands/connection_command.py:
```diff
@@ -29,7 +29,7 @@
 from airflow.hooks.base import BaseHook
 from airflow.models import Connection
-from airflow.secrets.local_filesystem import _create_connection, load_connections_dict
+from airflow.secrets.local_filesystem import load_connections_dict
```
What do you think to add this method as a static method to the Connection class - `Connection.from_dict`? I don't think CLI should depend on a backend secret.

mik-laj replied:
Ahhh. You don't want to change the visibility of this method, but you are using a different public method. We should therefore still create the `Connection.from_dict` method but as a separate PR.

natanweinberger replied:
Thanks for taking a look. I can add `Connection.from_dict`.

For the implementation, in order to avoid any dependencies on `secrets`, I'll need to move several helper functions. The existing function `load_connections_dict` accepts a filepath and relies on `_parse_secret_file`, which calls `_parse_[json,yaml,env]_file` to read the file contents and deserialize them.

I can either:
- move all these helper functions outside of `secrets/local_filesystem.py` (if so, into a utils file perhaps)
- duplicate the same functionality in the cli module (seems like a bad idea)
- create the method in the Connection class, but still import from `secrets` (doesn't eliminate the dependency)

What do you have in mind?

mik-laj replied:
Yes. We should avoid dependency on secrets. We already have a lot of recursive imports between various connection modules, secret and others, so creating new modules will cure those situations a bit.

> move all these helper functions outside of secrets/local_filesystem.py

I think it is the best solution.

natanweinberger replied:
Sounds good, I'll move this logic elsewhere. Do you still want that as a separate PR then or included in this one?

mik-laj replied:
You can do it in this PR.

natanweinberger replied:
This change is now implemented.

mik-laj commented on airflow/models/connection.py:
```diff
@@ -364,3 +364,14 @@ def get_connection_from_secrets(cls, conn_id: str) -> 'Connection':
+            e.g., {'conn_id': '', 'conn_type': '', 'login': '', ...}
+        :return: connection
+        """
+        return Connection(**conn_dict)
```
We have a function that has a similar purpose but is more complex and supports more input parameters. See: airflow.secrets.local_filesystem._create_connection  Can you move it here?

natanweinberger replied:
Sure, will do.

natanweinberger replied:
All set!

ashb commented on airflow/utils/parse.py:
```diff
@@ -0,0 +1,168 @@
+}
+
+
+def _parse_file(file_path: str) -> Dict[str, Any]:
```
Since this is used outside of this module it should be "non-private"

```suggestion
def parse_file(file_path: str) -> Dict[str, Any]:
```

natanweinberger replied:
✅

ashb commented on airflow/utils/parse.py:
```diff
@@ -0,0 +1,168 @@
+    if parse_errors:
+        raise AirflowFileParseException(
+            "Failed to load the file.", file_path=file_path, parse_errors=parse_errors
+        )
```
Just make the format specific parsers throw the errors directly?

natanweinberger replied:
No problem for me to do this, but heads up that this is not new code, just cut from `secrets/local_filesystem.py`. Not sure if there was a reason to do it this way when it was originally committed.

ashb commented on airflow/utils/parse.py:
```diff
@@ -0,0 +1,168 @@
+    except JSONDecodeError as e:
+        return {}, [FileSyntaxError(line_no=int(e.lineno), message=e.msg)]
+    if not isinstance(contents_dict, dict):
+        return {}, [FileSyntaxError(line_no=1, message="The file should contain the object.")]
```
```suggestion
        return {}, [FileSyntaxError(line_no=1, message="The file should contain an object.")]
```

natanweinberger replied:
✅

ashb commented on airflow/utils/parse.py:
```diff
@@ -0,0 +1,168 @@
+    for line_no, line in enumerate(content.splitlines(), 1):
+        if not line:
+            # Ignore empty line
+            continue
```
This loads the whole file in to memory, which is slightly wasteful. Instead we should do something like this:

```python
    with open(file_path) as f:
        for line_no, line in enumerate(f, 1):
            if not line.strip():
               continue
```

natanweinberger replied:
Same as above. Can do this, but it's existing code from `local_filesystem.py`. Just a heads up in case it had been originally committed this way intentionally, I don't have much context.

ashb replied:
> in case it had been originally committed this way intentionally

Very unlikely to have been intentional.

natanweinberger replied:
Ok, will make this change then!

natanweinberger replied:
I've been poking around on this and running into an issue with the existing tests.

The tests are using `unittest.mock.mock_open` to patch the call to `open` ([example](https://github.com/apache/airflow/blob/a73ba4b3af65ff392a05fe09f656a91b51401c4d/tests/secrets/test_local_filesystem.py#L36)). This works when you want to mock reading data from a file using `read()` or `readlines()`, which is the existing implementation.

However, `unittest.mock.mock_open` doesn't add compatibility for iterating over the lines of a file lazily until version 3.8. So, the `__iter__` method doesn't properly allow us to read the lines lazily in tests.

See here: https://docs.python.org/3/library/unittest.mock.html#mock-open
```
Changed in version 3.8: Added __iter__() to implementation so that iteration (such as in for loops) correctly consumes read_data.
```

Given that this is a minor enhancement, I'd suggest punting on it for the time being. It will expand the scope of this PR and should be done separately.

ashb commented on airflow/secrets/local_filesystem.py:
```diff
@@ -263,19 +71,32 @@ def load_connections_dict(file_path: str) -> Dict[str, Any]:
+            connection_by_conn_id[key] = Connection(uri=secret_values)
         else:
-            connection_by_conn_id[key] = _create_connection(key, secret_values)
+            raise AirflowException(f"Unexpected value type: {type(secret_values)}.")
```
```suggestion
            raise AirflowException(f"Unexpected value type: {type(secret_values).__name__}.")
```

natanweinberger replied:
✅

ashb commented on airflow/secrets/local_filesystem.py:
```diff
@@ -263,19 +71,32 @@ def load_connections_dict(file_path: str) -> Dict[str, Any]:
+                secret_values = secret_values[0]
+
+        if isinstance(secret_values, dict):
+            connection_by_conn_id[key] = Connection.from_dict(key, secret_values)
```
Do we need a method for this? Isn't it the same as

```suggestion
            connection_by_conn_id[key] = Connection(key, **secret_values)
```

natanweinberger replied:
Creating a method `Connection.from_dict` was a suggestion from @mik-laj: https://github.com/apache/airflow/pull/15425#discussion_r615929569

ashb replied:
Oh, this handles the extra to, which `**secret_values` wouldn't.

ashb reviewed (changes requested):
I'm not that happy about adding more files/modules under airflow.utils (as it's a bit of a dumping ground) but I guess it makes sense here.

ashb commented on airflow/models/connection.py:
```diff
@@ -377,3 +378,50 @@ def get_connection_from_secrets(cls, conn_id: str) -> 'Connection':
+        if not current_keys.issubset(connection_parameter_names):
+            illegal_keys = current_keys - connection_parameter_names
+            illegal_keys_list = ", ".join(sorted(illegal_keys))
+            raise AirflowException(
```
```suggestion
            raise TypeError(
```

natanweinberger replied:
Agreed with the same change on line 397, but this particular case shouldn't be a TypeError. It indicates that some keys in the dictionary aren't allowed. Maybe it can be a different type of exception though, what do you think?

ashb replied:
Oh, ValueError perhaps then?

natanweinberger replied:
Sounds good 👍

ashb commented on airflow/models/connection.py:
```diff
@@ -377,3 +378,50 @@ def get_connection_from_secrets(cls, conn_id: str) -> 'Connection':
+        """
+        if not isinstance(conn_dict, dict):
+            raise AirflowException(
+                f"Unexpected conn_dict type: {type(conn_dict)}. "
```
```suggestion
            raise TypeError(
                f"Unexpected conn_dict type: {type(conn_dict).__name__}. "
```

natanweinberger replied:
Implemented 👍

ashb commented on airflow/models/connection.py:
```diff
@@ -377,3 +378,50 @@ def get_connection_from_secrets(cls, conn_id: str) -> 'Connection':
+            illegal_keys = current_keys - connection_parameter_names
+            illegal_keys_list = ", ".join(sorted(illegal_keys))
+            raise AirflowException(
+                f"The object have illegal keys: {illegal_keys_list}. "
```
```suggestion
                f"The connection dict has illegal keys: {illegal_keys_list}. "
```

natanweinberger replied:
Implemented 👍

ashb reviewed (commented):
These are just some changes I had pending.

A full review coming on Monday

ashb commented on airflow/utils/parse.py:
```diff
@@ -0,0 +1,168 @@
+
+import airflow.utils.yaml as yaml
+from airflow.exceptions import AirflowException, AirflowFileParseException, FileSyntaxError
+from airflow.utils.file import COMMENT_PATTERN
```
```suggestion
from airflow.exceptions import AirflowException, AirflowFileParseException, FileSyntaxError
from airflow.utils import yaml
from airflow.utils.file import COMMENT_PATTERN
```

natanweinberger replied:
Got rid of utils/parse.py.

ashb commented on airflow/cli/commands/connection_command.py:
```diff
@@ -28,9 +28,9 @@
-from airflow.secrets.local_filesystem import _create_connection, load_connections_dict
 from airflow.utils import cli as cli_utils, yaml
 from airflow.utils.cli import suppress_logs_and_warning
+from airflow.utils.parse import parse_file
```
Hmmmm, I'm not really happy with this module name -- it doesn't _really_ tell me what its doing, not in specific enough terms.

We've already got `airflow.utils.file` that we could include this in -- I think I prefer that over a new file.

"parse_file" is also a bit ambigious, but I can't think of a better name.

natanweinberger replied:
Got rid of `utils/parse.py` and left the functionality where it was in `secrets/local_filesystem.py`.

ashb reviewed (changes requested):
Sorry for not getting around to this pre-2.1.

Why did we even move load_connections_dict from where it was? Your Vault PR doesn't need to use this parsing, so I think this PR would be simpler if we _only_ extract out the `_create_connection` function in to the `Connection.from_dict` and leave the file parsing where it was.

natanweinberger committed:
Add field 'extra' to Connection init

natanweinberger committed:
Fix connections import CLI

In connections_import, each connection was deserialized and stored into a
Connection model instance rather than a dictionary, so an erroneous call to the
dictionary methods .items() resulted in an AttributeError. With this fix,
connection information is loaded from dictionaries directly into the
Connection constructor and committed to the DB.

ashb commented on airflow/cli/commands/connection_command.py:
```diff
@@ -238,39 +238,24 @@ def connections_delete(args):
-            }
-            connection = _create_connection(conn_id, filtered_connection_values)
+            # Add the connection to the DB
+            connection = Connection(conn_id, **dict(conn_dict.items()))
```
```suggestion
            connection = Connection(conn_id, **conn_dict)
```

ashb commented on airflow/cli/commands/connection_command.py:
```diff
@@ -238,39 +238,24 @@ def connections_delete(args):
-                key: value for key, value in conn_values.items() if key in allowed_fields
-            }
-            connection = _create_connection(conn_id, filtered_connection_values)
+            # Add the connection to the DB
```
For compat we should add this

```suggestion
            if "extra_dejson" in conn_dict:
                conn_dict["extra"] = conn_dict.pop("extra_dejson")
            # Add the connection to the DB
```

ashb committed:
Apply suggestions from code review

ashb commented on airflow/cli/commands/connection_command.py:
```diff
@@ -238,39 +238,24 @@ def connections_delete(args):
-    """Helps import connections from a file"""
-    connections_dict = load_connections_dict(file_path)
+    """Load connections from a file and save them to the DB. On collision, skip."""
+    connections_dict = _parse_secret_file(file_path)
```
Wait, this shouldn't change, it should still call `load_connections_dict`, but we should change _that_ function to call.


_wait_. Looking at load_connections_dict, I see it is _already_ returning a dict with values of Connection objects!

Did this cli command actually work?!

natanweinberger replied:
Nope, I introduced a bug at the last minute. The values of the dict were indeed Connection objects, but I was treating them as dictionaries (and my test didn't catch it). I can change this back to `load_connections_dict`, but I thought this was more straightforward.

natanweinberger committed:
Use load_connections_dict in connections import
